File: EMS/EMS/EMSSessionsParser.m
//
//  EMSSessionsParser.m
//

#import "EMSSessionsParser.h"
#import "EMSSession.h"
#import "EMSSpeaker.h"

#import "CJCollection.h"
#import "CJItem.h"
#import "CJLink.h"


@interface EMSSession (JsonParser)

+ (EMSSession *) sessionWithItem:(CJItem *) item;

@end

@implementation EMSSession(JsonParser)

+ (EMSSession *)sessionWithItem:(CJItem *) item
{
    EMSSession *session = [[EMSSession alloc] init];
    
    session.keywords = nil;
    
    session.href = item.href;
    
    for (NSDictionary *dict in item.data) {
        
        NSString *field = dict[@"name"];
        NSObject *value = dict[@"value"];
        
        if ([@"format" isEqualToString:field]) {
            session.format = (NSString *) value;
        }
        if ([@"body" isEqualToString:field]) {
            session.body = (NSString *) value;
        }
        if ([@"state" isEqualToString:field]) {
            session.state = (NSString *) value;
        }
        if ([@"audience" isEqualToString:field]) {
            session.audience = (NSString *) value;
        }
        if ([@"title" isEqualToString:field]) {
            session.title = (NSString *) value;
        }
        if ([@"lang" isEqualToString:field]) {
            session.language = (NSString *) value;
        }
        if ([@"summary" isEqualToString:field]) {
            session.summary = (NSString *) value;
        }
        if ([@"level" isEqualToString:field]) {
            session.level = (NSString *) value;
        }
        if ([@"keywords" isEqualToString:field]) {
            session.keywords = [NSArray arrayWithArray:dict[@"array"]];
        }
    }
    
    NSMutableArray *speakers = [[NSMutableArray alloc] init];
    
    
    for (CJLink *link in item.links) {
        
        if ([@"alternate video" isEqualToString:link.rel]) {
            session.videoLink = link.href;
        }
        if ([@"attachment collection" isEqualToString:link.rel]) {
            session.attachmentCollection = link.href;
        }
        if ([@"speaker collection" isEqualToString:link.rel]) {
            session.speakerCollection = link.href;
        }
        if ([@"room item" isEqualToString:link.rel]) {
            session.roomItem = link.href;
        }
        if ([@"slot item" isEqualToString:link.rel]) {
            session.slotItem = link.href;
        }
        if ([@"speaker item" isEqualToString:link.rel]) {
            EMSSpeaker *speaker = [[EMSSpeaker alloc] init];
            
            speaker.href = link.href;
            speaker.name = link.prompt;
            
            [speakers addObject:speaker];
        }
    }
    
    session.speakers = [NSArray arrayWithArray:speakers];
    
    return session;
}
@end

@implementation EMSSessionsParser

- (NSArray *)processData:(NSData *)data forHref:(NSURL *)href error:(NSError **)error {
    NSError *parseError = nil;

    CJCollection *collection = [CJCollection collectionForNSData:data error:&parseError];

    if (!collection) {
        EMS_LOG(@"Failed to retrieve sessions %@ - %@ - %@", href, parseError, [parseError userInfo]);

        *error = parseError;

        return [NSArray array];
    }

    NSMutableArray *temp = [[NSMutableArray alloc] init];
    
    for (CJItem *item in collection.items) {
        EMSSession *session = [EMSSession sessionWithItem:item];
        [temp addObject:session];
    }

    return [NSArray arrayWithArray:temp];
}

- (void)parseData:(NSData *)responseData forHref:(NSURL *)href {
    NSError *error;

    NSArray *collection = [self processData:responseData forHref:href error:&error];

    [self.delegate finishedSessions:collection forHref:href error:error];
}

@end

File: EMS/Network/Parsers/EMSSessionsParser.swift
import Foundation
import CocoaLumberjack

@objc protocol EMSSessionsParserDelegate {
    optional func finishedSessions(sessions: [EMSSession], forHref href:NSURL, error:NSError?)
}

@objc class EMSSessionsParser : NSObject {
    var delegate : EMSSessionsParserDelegate?
    
    func processData(data: NSData, forHref href:NSURL) throws -> [EMSSession] {
        
        do {
            let collection = try CJCollection(forNSData:data)
            
            return collection.items.map({ item in
                let session = EMSSession()
                
                session.href = item.href
                
                for datum in item.data {
                    if let field = datum["name"] as? String {
                        if let value = datum["value"] as? String {
                            switch (field) {
                            case "format":
                                session.format = value
                            case "body":
                                session.body = value
                            case "state":
                                session.state = value
                            case "audience":
                                session.audience = value
                            case "title":
                                session.title = value
                            case "lang":
                                session.language = value
                            case "summary":
                                session.summary = value
                            case "level":
                                session.level = value
                            default:
                                DDLogDebug("Ignored string field \(field) for session")
                            }
                        }
                        
                        if let value = datum["array"] as? [String] {
                            switch (field) {
                            case "keywords":
                                session.keywords = value
                            default:
                                DDLogDebug("Ignored [string] field \(field) for session")
                            }
                        }
                    }
                }
                
                var speakers : [EMSSpeaker] = []
                
                for link in item.links {
                    switch (link.rel) {
                    case "alternate video":
                        session.videoLink = link.href
                    case "alternate":
                        session.link = link.href
                    case "attachment collection":
                        session.attachmentCollection = link.href
                    case "speaker collection":
                        session.speakerCollection = link.href
                    case "room item":
                        session.roomItem = link.href
                    case "slot item":
                        session.slotItem = link.href
                    case "speaker item":
                        let speaker = EMSSpeaker()
                        
                        speaker.href = link.href
                        speaker.name = link.prompt
                        
                        speakers.append(speaker)
                    default:
                        DDLogDebug("Ignored rel \(link.rel) for session")
                    }
                }
                
                session.speakers = speakers
                
                return session
            })
        } catch let error as NSError {
            DDLogError("Failed to retrieve sessions \(href) - \(error)")
            
            throw error
        }
    }
    
    func parseData(data: NSData, forHref url: NSURL) {
        do {
            let collection = try self.processData(data, forHref: url)
            
            self.delegate?.finishedSessions?(collection, forHref: url, error: nil)
        } catch let error as NSError {
            self.delegate?.finishedSessions?([], forHref: url, error: error)
        }
    }
}
